tests/system/Database/Live/UpsertTest.php

```diff
@@ -0,0 +1,393 @@
+<?php
+
+/**
+ * This file is part of CodeIgniter 4 framework.
+ *
+ * (c) CodeIgniter Foundation <admin@codeigniter.com>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\Database\Live;
+
+use CodeIgniter\Database\Exceptions\DatabaseException;
+use CodeIgniter\Test\CIUnitTestCase;
+use CodeIgniter\Test\DatabaseTestTrait;
+use stdclass;
+use Tests\Support\Database\Seeds\CITestSeeder;
+
+/**
+ * @group DatabaseLive
+ *
+ * @internal
+ */
+final class UpsertTest extends CIUnitTestCase
+{
+    use DatabaseTestTrait;
+
+    protected $refresh = true;
+    protected $seed    = CITestSeeder::class;
+
+    public function testUpsertOnUniqueIndex()
+    {
+        $userData = [
+            'email'   => 'upsertone@test.com',
+            'name'    => 'Upsert One',
+            'country' => 'US',
+        ];
+
+        $this->db->table('user')->upsert($userData);
+
+        $this->seeInDatabase('user', ['name' => 'Upsert One']);
+    }
+
+    public function testUpsertAndUpsertBatchWithObject()
+    {
+        $data = [];
+
+        $row          = new stdclass();
+        $row->name    = 'Pedro';
+        $row->email   = 'pedro@acme.com';
+        $row->country = 'El Salvador';
+        $data[]       = $row;
+
+        $row          = new stdclass();
+        $row->name    = 'Derek Jones';
+        $row->email   = 'derek@world.com';
+        $row->country = 'Canada';
+        $data[]       = $row;
+
+        $this->db->table('user')->upsertBatch($data);
+
+        $this->seeInDatabase('user', ['name' => 'Pedro']);
+        $this->seeInDatabase('user', ['name' => 'Derek Jones', 'country' => 'Canada']);
+
+        $row->country = 'Spain';
+        $this->db->table('user')->upsert($row);
+
+        $this->seeInDatabase('user', ['name' => 'Derek Jones', 'country' => 'Spain']);
+    }
+
+    public function testUpsertChangePrimaryKeyOnUniqueIndex()
+    {
+        $userData = [
+            'id'      => 5,
+            'email'   => 'ahmadinejad@world.com',
+            'name'    => 'Ahmadinejad',
+            'country' => 'Iran',
+        ];
+
+        // some databases don't allow update of auto incrememnt
+        if ($this->db->DBDriver === 'SQLSRV') {
+            $this->assertTrue(true);
+        } else {
+            $this->db->table('user')
+                ->onConstraint('email')
+                ->updateFields('id, name, country')
+                ->upsert($userData);
+
+            $new = $this->db->table('user')
+                ->getwhere(['email' => 'ahmadinejad@world.com'])
+                ->getRow();
+
+            $this->assertSame(5, (int) $new->id);
+        }
+    }
+
+    public function testGetCompiledUpsert()
```

Comment:
This test method is very long, and it will be longer if another DB driver is added.

I prefer to split into DB driver specific test case classes like
https://github.com/codeigniter4/CodeIgniter4/blob/develop/tests/system/Database/Live/OCI8/LastInsertIDTest.php

Reply:
From the developer perspective its nicer to have all the upsert tests in one place. This way when working on the code you only have to run one test file. 

This pattern is used many places:
tests/system/Database/Live/ForgeTest.php line 174
tests/system/Database/Live/ForgeTest.php line 434
tests/system/Database/Live/ForgeTest.php line 497
tests/system/Database/Live/ForgeTest.php line 825
tests/system/Database/Live/ForgeTest.php line 908

I have reorganized the code for this method so its very easy to read.

It is also nice to see them all together so you can see the differences. This makes refactoring easier.

I really find this method the best, but if you still find this unacceptable I will put them in separate files.

Reply:
Okay, you refactored this method, and it is easier to read than before.
Using `switch` is better than a lot of `if` like in ForgeTest.

Personally the ForgeTest is really difficult to read, so I would like to refactor.